api/v1alpha1/backend.go

```diff
@@ -2,147 +2,149 @@ package v1alpha1
 
 import (
 	"fmt"
+	"reflect"
+
 	"github.com/vshn/k8up/cfg"
 
 	corev1 "k8s.io/api/core/v1"
 )
 
-type Backend struct {
-	// RepoPasswordSecretRef references a secret key to look up the restic repository password
-	RepoPasswordSecretRef *corev1.SecretKeySelector `json:"repoPasswordSecretRef,omitempty"`
-	Local                 *LocalSpec                `json:"local,omitempty"`
-	S3                    *S3Spec                   `json:"s3,omitempty"`
-	GCS                   *GCSSpec                  `json:"gcs,omitempty"`
-	Azure                 *AzureSpec                `json:"azure,omitempty"`
-	Swift                 *SwiftSpec                `json:"swift,omitempty"`
-	B2                    *B2Spec                   `json:"b2,omitempty"`
-	Rest                  *RestServerSpec           `json:"rest,omitempty"`
-}
-
-// GetCredentialEnv will return a map containing the correct
-func (b *Backend) GetCredentialEnv() map[string]*corev1.EnvVarSource {
-	vars := make(map[string]*corev1.EnvVarSource)
-
-	if b.RepoPasswordSecretRef != nil {
-		vars[cfg.ResticPasswordEnvName] = &corev1.EnvVarSource{
-			SecretKeyRef: b.RepoPasswordSecretRef,
-		}
-	}
-
-	if b.Azure != nil {
-		return b.Azure.EnvVars(vars)
-	}
-
-	if b.B2 != nil {
-		return b.B2.EnvVars(vars)
+type (
+	// Backend allows configuring several backend implementations.
+	// It is expected that users only configure one storage type.
+	Backend struct {
+		// RepoPasswordSecretRef references a secret key to look up the restic repository password
+		RepoPasswordSecretRef *corev1.SecretKeySelector `json:"repoPasswordSecretRef,omitempty"`
+		Local                 *LocalSpec                `json:"local,omitempty"`
+		S3                    *S3Spec                   `json:"s3,omitempty"`
+		GCS                   *GCSSpec                  `json:"gcs,omitempty"`
+		Azure                 *AzureSpec                `json:"azure,omitempty"`
+		Swift                 *SwiftSpec                `json:"swift,omitempty"`
+		B2                    *B2Spec                   `json:"b2,omitempty"`
+		Rest                  *RestServerSpec           `json:"rest,omitempty"`
 	}
 
-	if b.GCS != nil {
-		return b.GCS.EnvVars(vars)
-	}
+	// +k8s:deepcopy-gen=false
 
-	if b.Local != nil {
-		return b.Local.EnvVars(vars)
+	// BackendInterface represents a Backend for internal use.
+	BackendInterface interface {
+		EnvVars(vars map[string]*corev1.EnvVarSource) map[string]*corev1.EnvVarSource
+		GetRepositoryString() string
```

Comment:
Any objections in simply calling this `String()`? Then the implementations would also implement the Stringer interface.

Reply:
Hm, personally I find `String()` to be a very vague name as to what to expect. But I see the argument for the Stringer interface.